eth/db/journal.py

```diff
@@ -24,237 +26,258 @@ class DeletedEntry:
 # 1. key in wrapped
 # 2. key modified in journal
 # 3. key deleted
-DELETED_ENTRY = DeletedEntry()
+DELETE_WRAPPED = DeletedEntry()
 
 # 1. key not in wrapped
 # 2. key created in journal
 # 3. key deleted
-ERASE_CREATED_ENTRY = DeletedEntry()
+REVERT_TO_WRAPPED = DeletedEntry()
+
+ChangesetValue = Union[bytes, DeletedEntry]
+ChangesetDict = Dict[bytes, ChangesetValue]
+
+get_next_checkpoint_id = cast(Callable[[], JournalDBCheckpoint], count().__next__)
 
 
 class Journal(BaseDB):
     """
     A Journal is an ordered list of changesets.  A changeset is a dictionary
-    of database keys and values.  The values are tracked changes that were
-    written after the changeset was created
+    of database keys and values.  The values are tracked changes which give
+    the information needed to revert a changeset to an old checkpoint.
 
-    Changesets are referenced by a random uuid4.
+    Changesets are referenced by an internally-generated integer. This is *not* threadsafe.
     """
+    __slots__ = [
+        '_journal_data',
+        '_clears_at',
+        '_current_values',
+        '_ignore_wrapped_db',
+        '_checkpoint_stack',
```

Comment:
Not for this PR but maybe worth checking out something like https://github.com/cjrh/autoslot
If we use `__slots__` at more places I would feel increasingly uneasy about having to keep this "stringly typed" array in sync with the actual attributes.